rust/connlib/tunnel/src/sockets.rs

```diff
@@ -13,10 +13,47 @@ use crate::Result;
 pub struct Sockets {
     socket_v4: Option<Socket>,
     socket_v6: Option<Socket>,
+
+    #[cfg(unix)]
+    protect: Box<dyn Fn(std::os::fd::RawFd) -> io::Result<()> + Send + 'static>,
```

Comment:
I'm not that sure about this platform-dependent callback.

What if instead of having a `protect` callback, we have a `create_socket` callback, that returns the socket that each platform can create as it want.

Another idea, that perhaps works better, is getting a `socket2::UdpSocket` instead of a `tokio::Socket`and keep the same logic as before where the client pass the socket. Those can be created outside a tokio context and can be converted into `tokio::Socket`

Reply:
@conectado Would it make sense to save this refactor for another PR just to get these fixes into `main`? I believe @thomaseizinger is out this week and you next week so maybe sometime after that.

Reply:
https://github.com/firezone/firezone/issues/4314

Reply:
yeah, I'll be in favor of this (sorry, I missed this comment) fixing main asap seems more important :)

Reply:
I think that is a great idea @conectado ! Despite some duplication across the clients, I think it is much cleaner to let them choose, how to initialize they sockets. There is actually no reason why `connlib` should dictate the use of a random port. it should be up to the client to pick the port. That would make it trivial to e.g. implement a `PORT` env variable on the gateway.

Reply:
Very cool 😎